src/main/java/org/broadinstitute/hellbender/utils/mcmc/MinibatchSliceSampler.java

```diff
@@ -0,0 +1,228 @@
+package org.broadinstitute.hellbender.utils.mcmc;
+
+
+import org.apache.commons.math3.distribution.TDistribution;
+import org.apache.commons.math3.primes.Primes;
+import org.apache.commons.math3.random.RandomGenerator;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.param.ParamUtils;
+
+import java.util.*;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Implements slice sampling of a continuous, univariate, unnormalized probability density function (PDF),
+ * which is assumed to be unimodal.  See Neal 2003 at https://projecteuclid.org/euclid.aos/1056562461 for details.
+ * Minibatching is implemented as in Dubois et al. 2014 at http://proceedings.mlr.press/v33/dubois14.pdf and requires
+ * that the PDF, which is assumed to be a posterior function of a parameter value and the data, is specified in terms
+ * of a prior, a likelihood, and the data.
+ *
+ * Note that likelihoods are cached and retrieved based on the hash of {@code DATA}, i.e., we assume it is
+ * cheaper to compute this hash rather than the likelihood.
+ *
+ * @author Samuel Lee &lt;slee@broadinstitute.org&gt;
+ */
+public final class MinibatchSliceSampler<DATA> extends AbstractSliceSampler {
+    private final List<DATA> data;
+    private final Function<Double, Double> logPrior;
+    private final BiFunction<DATA, Double, Double> logLikelihood;
+    private final Integer minibatchSize;
+    private final Double approxThreshold;
+
+    private final int numDataPoints;
+
+    private Double xSampleCache = null;
+    private Double logPriorCache = null;
+    private Map<DATA, Double> logLikelihoodsCache = null;    //data -> log likelihood
+
+    /**
+     * Creates a new sampler for a bounded univariate random variable, given a random number generator, a list of data,
+     * a continuous, univariate, unimodal, unnormalized log probability density function
+     * (assumed to be a posterior and specified by a prior and a likelihood),
+     * hard limits on the random variable, a step width, a minibatch size, and a minibatch approximation threshold.
+     * @param rng                       random number generator
```

Comment:
Can you add the usual `Never {@code null}` for parameters that are not allowed to be null.